src/control/cmd/daos/container.go

```diff
@@ -281,102 +281,242 @@ func (cmd *containerCreateCmd) Execute(_ []string) (err error) {
 	}
 	defer disconnectPool()
 
-	ap.c_op = C.CONT_CREATE
+	var contID string
+	if cmd.Path != "" {
+		contID, err = cmd.contCreateUNS()
+	} else {
+		contID, err = cmd.contCreate()
+	}
+	if err != nil {
+		return err
+	}
 
-	if cmd.Group != "" {
-		ap.group = C.CString(cmd.Group.String())
-		defer freeString(ap.group)
+	if err := cmd.openContainer(C.DAOS_COO_RO); err != nil {
+		return errors.Wrapf(err, "failed to open new container %s", contID)
 	}
-	if cmd.ACLFile != "" {
-		ap.aclfile = C.CString(cmd.ACLFile)
-		defer freeString(ap.aclfile)
+	defer cmd.closeContainer()
+
+	var ci *containerInfo
+	ci, err = cmd.queryContainer()
+	if err != nil {
+		if errors.Cause(err) != daos.NoPermission {
+			return errors.Wrapf(err, "failed to query new container %s", contID)
+		}
+
+		// Special case for creating a container without permission to query it.
+		cmd.Errorf("container %s was created, but query failed", contID)
+
+		ci = new(containerInfo)
+		ci.PoolUUID = &cmd.poolUUID
+		ci.Type = cmd.Type.String()
+		ci.ContainerUUID = &cmd.contUUID
+		ci.ContainerLabel = cmd.Args.Label
 	}
 
-	ap._type = cmd.Type.Type
-	switch ap._type {
-	case C.DAOS_PROP_CO_LAYOUT_POSIX:
+	if cmd.jsonOutputEnabled() {
+		return cmd.outputJSON(ci, nil)
+	}
+
+	var bld strings.Builder
+	if err := printContainerInfo(&bld, ci, false); err != nil {
+		return err
+	}
+	cmd.Info(bld.String())
+
+	return nil
+}
+
+func (cmd *containerCreateCmd) contCreate() (string, error) {
+	props, cleanupProps, err := cmd.getCreateProps()
+	if err != nil {
+		return "", err
+	}
+	defer cleanupProps()
+
+	var rc C.int
+	var contUUID C.uuid_t
+	if cmd.Type.Type == C.DAOS_PROP_CO_LAYOUT_POSIX {
+		var attr C.dfs_attr_t
+
 		// POSIX containers have extra attributes
 		if cmd.ChunkSize.Set {
-			ap.chunk_size = cmd.ChunkSize.Size
+			attr.da_chunk_size = cmd.ChunkSize.Size
 		}
 		if cmd.ObjectClass.Set {
-			ap.oclass = cmd.ObjectClass.Class
+			attr.da_oclass_id = cmd.ObjectClass.Class
 		}
 		if cmd.DirObjectClass.Set {
-			ap.dir_oclass = cmd.DirObjectClass.Class
+			attr.da_dir_oclass_id = cmd.DirObjectClass.Class
 		}
 		if cmd.FileObjectClass.Set {
-			ap.file_oclass = cmd.FileObjectClass.Class
+			attr.da_file_oclass_id = cmd.FileObjectClass.Class
 		}
 		if cmd.Mode.Set {
-			ap.mode = cmd.Mode.Mode
+			attr.da_mode = cmd.Mode.Mode
 		}
 		if cmd.CHints != "" {
-			ap.hints = C.CString(cmd.CHints)
-			defer freeString(ap.hints)
+			hint := C.CString(cmd.CHints)
+			defer freeString(hint)
+			C.strncpy(&attr.da_hints[0], hint, C.DAOS_CONT_HINT_MAX_LEN-1)
```

Comment:
Output of `strncpy()` call should be checked. From what I understand, If it is equal to ` C.DAOS_CONT_HINT_MAX_LEN-1`, the input string will be truncated and could be not null-terminated.
Same remark for the function `contCreateUNS()`.

Reply:
Go should be automatically zeroing structures, so I would expect the da_hints field to be all zeroes. If the input gets truncated, the last byte will be a zero.

Reply:
Sounds logical for the zero padding.
NIT, do you think some Notice message should be done to indicate that the hint string has been truncated ?